gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java

```diff
@@ -55,69 +67,201 @@ public DefaultSubscriptionDispatcher(
+
+                    return subscription;
                 }
+            } catch (Exception ex) {
+                LOGGER.error("Unable to dispatch subscription id[{}] api[{}]", subscription.getId(), subscription.getApi(), ex);
             }
-        } else {
-            actives.computeIfPresent(
-                subscription.getId(),
-                (s, disposable) -> {
-                    disposable.dispose();
-                    // Return null to remove the item from the map
-                    return null;
+        }
+        return null;
+    }
+
+    private Completable buildSubscriptionObservable(Subscription subscription, ApiReactor reactorHandler, MutableExecutionContext context) {
+        return reactorHandler
+            .handle(context)
+            // Apply a delay before starting the subscription if it has a starting date
+            .compose(delayToStartDate(subscription))
+            // Apply a timeout to the subscription if it has an ending date
+            .compose(timeoutAtEndingDate(subscription))
+            // Manage functional errors, for example when a subscription expires
+            .onErrorComplete(manageErrors())
+            // In case of unhandled exception, retry ON_SUBSCRIPTION_ERROR_RETRY_COUNT times with a delay of ON_SUBSCRIPTION_ERROR_RETRY_DELAY_MS between attempts
+            .compose(RxHelper.retry(ON_SUBSCRIPTION_ERROR_RETRY_COUNT, ON_SUBSCRIPTION_ERROR_RETRY_DELAY_MS, MILLISECONDS))
+            .onErrorResumeNext(
+                t -> {
+                    // Here, manage ERROR status for subscription and send command to repository
+                    return Completable.complete();
                 }
             );
+    }
+
+    private CompletableTransformer delayToStartDate(Subscription subscription) {
+        return upstream -> {
+            if (subscription.getStartingAt() != null) {
+                long subscriptionStartDelay = getMillisecondsTo(subscription.getStartingAt());
+                return upstream.delaySubscription(subscriptionStartDelay, MILLISECONDS);
+            }
+            return upstream;
+        };
+    }
+
+    private CompletableTransformer timeoutAtEndingDate(Subscription subscription) {
+        return upstream -> {
+            if (subscription.getEndingAt() != null) {
+                long subscriptionEndDelay = getMillisecondsTo(subscription.getEndingAt());
+                return upstream.timeout(
+                    subscriptionEndDelay,
+                    MILLISECONDS,
+                    Completable.error(new SubscriptionExpiredException(subscription.getId()))
+                );
+            }
+            return upstream;
+        };
+    }
+
+    private static Predicate<Throwable> manageErrors() {
+        return throwable -> {
+            if (throwable instanceof SubscriptionExpiredException) {
+                LOGGER.info(throwable.getMessage());
```

Comment:
if the throwable message is something like `Subscription [SUB-ID] disabled because it has expired at [2022-11-25 07:44]`, is it more convenient to you ?

Reply:
Perhaps opt for a debug level as the subscription expiration is a functional case or just simply not log this. I don't see any value to log it in the server's log. We should not log functional events like this, especially if there is no action required because "it's normal" and the api publisher will certainly not check the server logs (in a real life environment).  Going that path, why not doing it for everything: a request that returns a 400 http code, a policy condition that returns false, ... 😜